File: VerbumKit/Sources/Models/AuthSession.swift
import Foundation

/// Firebase owns credential persistence. Views receive only this safe snapshot.
public struct AuthSession: Equatable, Sendable {
    public let id: String
    public let email: String?
    public let isAnonymous: Bool
    public let isEmailVerified: Bool
    public let displayName: String?
    public let createdAt: Date?
    public let providers: [String]
    public var hasPassword: Bool { providers.contains("password") }
    public init(id: String, email: String?, isAnonymous: Bool, isEmailVerified: Bool,
                displayName: String? = nil, createdAt: Date? = nil, providers: [String] = []) {
        self.id = id; self.email = email; self.isAnonymous = isAnonymous; self.isEmailVerified = isEmailVerified
        self.displayName = displayName; self.createdAt = createdAt; self.providers = providers
    }
}

public enum AccountFailure: Error, Equatable, Sendable {
    case invalidEmail, passwordRequired, passwordTooShort, passwordMismatch, nameRequired
    case credentials, emailInUse, network, tooManyRequests, disabled, configuration, recentLoginRequired, unexpected
}

public enum AccountValidation {
    public static func email(_ value: String) -> String { value.trimmingCharacters(in: .whitespacesAndNewlines) }
    public static func validate(email: String, password: String, confirmation: String?, reset: Bool = false) -> AccountFailure? {
        let address = self.email(email)
        guard address.count <= 254, address.range(of: #"^[^\s@]+@[^\s@]+\.[^\s@]+$"#, options: .regularExpression) != nil else { return .invalidEmail }
        if reset { return nil }
        guard !password.isEmpty else { return .passwordRequired }
        if let confirmation {
            guard password.unicodeScalars.count >= 8 else { return .passwordTooShort }
            guard password == confirmation else { return .passwordMismatch }
        }
        return nil
    }
}

File: android/core/models/src/main/kotlin/com/nexussoft/verbum/models/AuthSession.kt
package com.nexussoft.verbum.models

data class AuthSession(val id: String, val email: String?, val isAnonymous: Boolean, val isEmailVerified: Boolean,
    val displayName: String? = null, val createdAt: Long? = null, val providers: List<String> = emptyList(),
) {
    val hasPassword: Boolean get() = "password" in providers
    val initials: String? get() {
        val parts = displayName?.trim()?.split(Regex("\\s+"))?.filter { it.isNotEmpty() }.orEmpty()
        return if (parts.isNotEmpty()) (parts.first().take(1) + if (parts.size > 1) parts.last().take(1) else "").uppercase()
            else email?.take(1)?.uppercase()
    }
}

enum class AccountFailure {
    invalidEmail, nameRequired, passwordRequired, passwordTooShort, passwordMismatch, credentials, emailInUse,
    network, tooManyRequests, disabled, configuration, recentLoginRequired, unexpected,
}
class AccountException(val failure: AccountFailure) : Exception(failure.name)

object AccountValidation {
    private val emailPattern = Regex("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$")
    fun email(value: String) = value.trim()
    fun validate(email: String, password: String, confirmation: String? = null, reset: Boolean = false): AccountFailure? {
        val address = email(email)
        if (address.length > 254 || !emailPattern.matches(address)) return AccountFailure.invalidEmail
        if (reset) return null
        if (password.isEmpty()) return AccountFailure.passwordRequired
        if (confirmation != null) {
            if (password.codePointCount(0, password.length) < 8) return AccountFailure.passwordTooShort
            if (password != confirmation) return AccountFailure.passwordMismatch
        }
        return null
    }
}
